crates/turborepo-lib/src/process/mod.rs

```diff
@@ -0,0 +1,285 @@
+//! `process`
+//!
+//! This module contains the code that is responsible for running the commands
+//! that are queued by run. It consists of a set of child processes that are
+//! spawned and managed by the manager. The manager is responsible for
+//! running these processes to completion, forwarding signals, and closing
+//! them when the manager is closed.
+//!
+//! As of now, the manager will execute futures in a random order, and
+//! must be either `wait`ed on or `stop`ped to drive state.
+
+mod child;
+
+use std::{
+    sync::{atomic::AtomicBool, Arc, Mutex},
+    time::Duration,
+};
+
+pub use child::Command;
+use tokio::task::JoinSet;
+use tracing::{debug, trace};
+
+/// A process manager that is responsible for spawning and managing child
+/// processes. When the manager is Open, new child processes can be spawned
+/// using `spawn`. When the manager is Closed, all currently-running children
+/// will be closed, and no new children can be spawned.
+#[derive(Debug, Clone)]
+pub struct ProcessManager {
+    is_closing: Arc<AtomicBool>,
+    children: Arc<Mutex<Vec<child::Child>>>,
+}
+
+impl ProcessManager {
+    pub fn new() -> Self {
+        ProcessManager {
+            is_closing: Arc::new(AtomicBool::new(false)),
+            children: Arc::new(Mutex::new(Vec::new())),
+        }
+    }
+}
+
+impl ProcessManager {
+    /// Spawn a new child process to run the given command.
+    ///
+    /// The handle of the child can be either waited or stopped by the caller,
+    /// as well as the entire process manager.
+    ///
+    /// If spawn returns None,
+    pub fn spawn(&self, command: child::Command, stop_timeout: Duration) -> Option<child::Child> {
+        if self.is_closing.load(std::sync::atomic::Ordering::Relaxed) {
+            return None;
+        }
+        let child = child::Child::spawn(command, child::ShutdownStyle::Graceful(stop_timeout));
+        {
+            let mut lock = self.children.lock().unwrap();
+            lock.push(child.clone());
+        }
+        Some(child)
+    }
+
+    /// Stop the process manager, closing all child processes. On posix
+    /// systems this will send a SIGINT, and on windows it will just kill
+    /// the process immediately.
+    pub async fn stop(&self) {
+        self.is_closing
+            .store(true, std::sync::atomic::Ordering::Relaxed);
```

Comment:
Do we need to be concerned with the behavior of multiple `stop` calls?

Reply:
No, this is intentional to support 'upgrading' the urgency in the event we started a wait but then decided to stop.